overlord/snapshotstate/backend/reader.go

```diff
@@ -0,0 +1,374 @@
+// in the snapshot) with that contained in the snapshot.
+func (r *Reader) Restore(ctx context.Context, usernames []string, logf Logf) error {
+	_, err := r.restore(ctx, false, usernames, logf)
+	return err
+}
+
+// RestoreLeavingTrash is like Restore, but it leaves the old data in
+// a trash directory.
+func (r *Reader) RestoreLeavingTrash(ctx context.Context, usernames []string, logf Logf) (*Trash, error) {
+	return r.restore(ctx, true, usernames, logf)
+}
+
+func (r *Reader) restore(ctx context.Context, leaveTrash bool, usernames []string, logf Logf) (b *Trash, e error) {
+	b = &Trash{}
+	defer func() {
+		if e != nil {
+			logger.Noticef("Restore of snapshot %q failed (%v); undoing.", r.Name(), e)
+			b.Revert()
+			b = nil
+		} else if !leaveTrash {
+			logger.Debugf("Restore of snapshot succeeded and no trash requested; cleaning up.")
+			b.Cleanup()
+			b = nil
+		}
+	}()
+
+	sort.Strings(usernames)
+	isRoot := sys.Geteuid() == 0
+	si := snap.MinimalPlaceInfo(r.Snap, r.Revision)
+	hasher := crypto.SHA3_384.New()
+	var sz sizer
+
+	for entry := range r.SHA3_384 {
+		if err := ctx.Err(); err != nil {
+			return b, err
+		}
+
+		var dest string
+		isUser := isUserArchive(entry)
+		uid := sys.UserID(osutil.NoChown)
+		gid := sys.GroupID(osutil.NoChown)
+
+		if !isUser {
+			if entry != archiveName {
+				// hmmm
+				logf("Skipping restore of unknown entry %q.", entry)
+				continue
+			}
+			dest = si.DataDir()
+		} else {
+			username := entryUsername(entry)
+			if len(usernames) > 0 && !strutil.SortedListContains(usernames, username) {
+				logger.Debugf("In restoring snapshot %q, skipping entry %q by user request.", r.Name(), username)
+				continue
+			}
+			usr, err := userLookup(username)
+			if err != nil {
+				logf("Skipping restore of user %q: %v.", username, err)
+				continue
+			}
+
+			dest = si.UserDataDir(usr.HomeDir)
+			fi, err := os.Stat(usr.HomeDir)
+			if err != nil {
+				if osutil.IsDirNotExist(err) {
+					logf("Skipping restore of %q as %q doesn't exist.", dest, usr.HomeDir)
+				} else {
+					logf("Skipping restore of %q: %v.", dest, err)
+				}
+				continue
+			}
+
+			if !fi.IsDir() {
+				logf("Skipping restore of %q as %q is not a directory.", dest, usr.HomeDir)
+				continue
+			}
+
+			if st, ok := fi.Sys().(*syscall.Stat_t); ok && isRoot {
+				// the mkdir below will use the uid/gid of usr.HomeDir
+				if st.Uid > 0 {
+					uid = sys.UserID(st.Uid)
+				}
+				if st.Gid > 0 {
+					gid = sys.GroupID(st.Gid)
+				}
+			}
+		}
+		parent, revdir := filepath.Split(dest)
+
+		exists, isDir, err := osutil.DirExists(parent)
+		if err != nil {
+			return b, err
+		}
+		if !exists {
+			// NOTE that the chown won't happen (it'll be NoChown)
+			// for the system path, and we won't be creating the
+			// user's home (as we skip restore in that case).
+			// Also no chown happens for root/root.
+			if err := osutil.MkdirAllChown(parent, 0755, uid, gid); err != nil {
```

Comment:
as the comment says, we explicitly check for and skip the restore in the case the home does not exist.
(see the `os.Stat(usr.HomeDir)` above)

Reply:
Missed that, thanks!

OTOH, should we check if SnapDataDir exists when restoring non-user data?

Reply:
we check if the parent of `si.DataDir()` (that is, `<SnapDataDir>/<snap>`) exists and create it (as root:root) if it doesn't. I think this is reasonable; `SnapDataDir` not existing might indicate a brand new snap (with the snapshot being 'imported' from elsewhere), which is probably fine.

Of course, `<SnapDataDir>/<snap>` might not exist because `<snap>` was `remove`d, but you just asked to restore its data so that's probably ok, at least in backend when dealing with snapshots and not snapshot sets -- whether the frontend should exclude non-installed snaps from being restored unless explicitly requested is fair (and open) question.